### ISSUE TITLE
Linear Regression

### ISSUE THREAD

**Issue by Wzz970205** [Wzz970205/qlib_example#1]
import sys, site
from pathlib import Path
import qlib
import pandas as pd
from qlib.constant import REG_CN
from utilities import prepare_xy, fillna_with_median
from qlib.contrib.model.gbdt import LGBModel
from qlib.contrib.data.handler import Alpha158
from qlib.contrib.strategy.signal_strategy import TopkDropoutStrategy
from qlib.contrib.evaluate import (backtest_daily as normal_backtest, risk_analysis)
from qlib.utils import exists_qlib_data, init_instance_by_config
from qlib.workflow import R
from qlib.workflow.record_temp import SignalRecord, PortAnaRecord
from qlib.utils import flatten_dict
from qlib.data.dataset.handler import DataHandlerLP
from sklearn.linear_model import LinearRegression
import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset
import networkx as nx
import torch_geometric
from torch_geometric.nn import GCNConv
from torch_geometric.data import Data as GeoData

if __name__ == '__main__':
    from multiprocessing import freeze_support
    freeze_support()

    scripts_dir = Path.cwd().parent.joinpath("quant\scripts")
    if not scripts_dir.joinpath("get_data.py").exists():
        # download get_data.py script
        scripts_dir.mkdir(parents=True, exist_ok=True)
        import requests

        with requests.get("https://raw.githubusercontent.com/microsoft/qlib/main/scripts/get_data.py", timeout=10) as resp:
            with open(scripts_dir.joinpath("get_data.py"), "wb") as fp:
                fp.write(resp.content)

    #####download_data
    provider_uri = "E:/HFIT/quant/qlib_data/"  # target_dir
    if not exists_qlib_data(provider_uri):
        print(f"Qlib data is not found in {provider_uri}")
        sys.path.append(str(scripts_dir))
        from scripts.get_data import GetData

        GetData().qlib_data(target_dir=provider_uri, region=REG_CN)
    qlib.init(provider_uri=provider_uri, region=REG_CN)

    ###define config
    market = "csi300"
    benchmark = "SH000300"

    data_handler_config = {
        "start_time": "2008-01-01",
        "end_time": "2020-08-01",
        "fit_start_time": "2008-01-01",
        "fit_end_time": "2014-12-31",
        "instruments": market,
    }

    ###
    #LinearModel

    task = {
        "model": {
            "class": "LinearModel",
            "module_path": "qlib.contrib.model.linear",
            "kwargs": {
                "estimator": "ols",
            },
        },
        "dataset": {
            "class": "DatasetH",
            "module_path": "qlib.data.dataset",
            "kwargs": {
                "handler": {
                    "class": "Alpha158",
                    "module_path": "qlib.contrib.data.handler",
                    "kwargs": data_handler_config,
                },
                "segments": {
                    "train": ("2008-01-01", "2014-12-31"),
                    "valid": ("2015-01-01", "2016-12-31"),
                    "test": ("2017-01-01", "2020-08-01"),
                },
            },
        },
    }

    # model initialization
    dataset = init_instance_by_config(task["dataset"])
    df_train = dataset.prepare("train")
    df_valid = dataset.prepare("valid")
    df_test = dataset.prepare("test")

    X_train, y_train = prepare_xy(df_train)
    X_valid, y_valid = prepare_xy(df_valid)
    X_test, y_test = prepare_xy(df_test)

    ###############OLS模型
    lr_model = LinearRegression()
    lr_model.fit(X_train, y_train)
    pred_lr = lr_model.predict(X_test)

    #####LSTM 模型
    X_train_lstm = torch.tensor(X_train[:, None, :], dtype=torch.float32)  # (batch, seq=1, feature)
    y_train_lstm = torch.tensor(y_train, dtype=torch.float32)

    model_lstm = LSTMModel(input_size=X_train.shape[1])
    optimizer = torch.optim.Adam(model_lstm.parameters(), lr=1e-3)
    loss_fn = nn.MSELoss()

    for epoch in range(5):
        optimizer.zero_grad()
        output = model_lstm(X_train_lstm).squeeze()
        loss = loss_fn(output, y_train_lstm)
        loss.backward()
        optimizer.step()

    X_test_lstm = torch.tensor(X_test[:, None, :], dtype=torch.float32)
    pred_lstm = model_lstm(X_test_lstm).detach().numpy().flatten()
    print("test!")
